agents/web-api/src/main/java/edu/jhuapl/dorset/agents/StockAgent.java

```diff
@@ -0,0 +1,274 @@
+import edu.jhuapl.dorset.http.HttpClient;
+
+public class StockAgent extends AbstractAgent {
+
+    private static final Logger logger = LoggerFactory.getLogger(StockAgent.class);
+
+    private static final String SUMMARY = "Stock ticker that gives historical stock information for NASDAQ and NYSE companies";
+    private static final String [] EXAMPLE = new String[] {"Stocks Facebook", "Facebook"};
+
+    private static final CellProcessor[] processors = new CellProcessor[] {
+            new NotNull(), new NotNull(), new NotNull(), new Optional(),
+            new Optional(), new Optional(), new NotNull(), new Optional() };
+    
+    private String baseurl = "https://www.quandl.com/api/v3/datasets/WIKI/";
+    private String apiKey;
+    private HttpClient client;
+    private TreeMap<String, CompanyInfo> stockSymbolMap;
+    private static final int DAYS_IN_A_MONTH = 30;
+
+
+    /**
+     * Stock agent
+     *
+     * The Quandl API provides stock information for various companies world
+     * wide. The StockAgent uses said API to scrape historical closing prices of
+     * NASDAQ and NYSE companies.
+     *
+     * @param client An http client object
+     * @param apiKey A Quandl API key
+     */
+    public StockAgent(HttpClient client, String apiKey) {
+        this.client = client;
+        this.apiKey = apiKey;
+
+        this.stockSymbolMap = new TreeMap<String, CompanyInfo>(String.CASE_INSENSITIVE_ORDER);
+        csvFileReader("stockagent/NASDAQ_Companies.csv");
+        csvFileReader("stockagent/NYSE_Companies.csv");
+
+        this.setDescription(new Description("stock ticker", SUMMARY, EXAMPLE));
+
+    }
+
+    @Override
+    public AgentResponse process(AgentRequest request) {
+        logger.debug("Handling the request: " + request.getText());
+
+        // remove trigger word "stocks"
+        String regex = "\\bstocks\\b";
+        Pattern pat = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
+        String requestCompanyName = pat.matcher(request.getText()).replaceAll("").trim();
+
+        CompanyInfo stockCompanyInfo = findStockSymbol(requestCompanyName);
+        String keywordCompanyName = null;
+        String keywordCompanySymbol = null;
+
+        if (stockCompanyInfo != null) {
+            keywordCompanyName = stockCompanyInfo.getName();
+            keywordCompanySymbol = stockCompanyInfo.getSymbol();
+        }
+
+        String json = null;
+
+        if (keywordCompanySymbol == null) {
+            return new AgentResponse(new ResponseStatus(ResponseStatus.Code.AGENT_DID_NOT_KNOW_ANSWER,
+                            "I am sorry, I don't understand which company you are asking about."));
+        }
+
+        json = requestData(keywordCompanySymbol);
+
+        if (json == null) {
+
+            // replace ".." with "." to maintain proper grammar when the
+            // keyword contains an abbreviation
+            if (keywordCompanyName != null) {
+                return new AgentResponse(
```

Comment:
Is name ever null when symbol is not null? If not, an error response has already been sent back earlier